ProjFS.Mac/PrjFSKext/PrjFSProviderUserClient.hpp

```diff
@@ -1,73 +1,9 @@
 #pragma once
 
 #include "PrjFSClasses.hpp"
-#include "Locks.hpp"
-#include "public/Message.h"
-#include "VirtualizationRoots.hpp"
-#include <IOKit/IOUserClient.h>
+#include <stdint.h>
 
-struct MessageHeader;
-struct VirtualizationRoot;
-class IOSharedDataQueue;
-class PrjFSProviderUserClient : public IOUserClient
-{
-    OSDeclareDefaultStructors(PrjFSProviderUserClient);
-private:
-    typedef IOUserClient super;
-    IOSharedDataQueue* dataQueue;
-    IOMemoryDescriptor* dataQueueMemory;
-    Mutex dataQueueWriterMutex;
-    pid_t pid;
-    // The root for which this is the provider; RootHandle_None prior to registration
-    VirtualizationRootHandle virtualizationRootHandle;
-
-    static const IOExternalMethodDispatch ProviderUserClientDispatch[];
-
-public:
-    // IOUserClient methods:
-    virtual bool initWithTask(
-        task_t owningTask, void* securityToken, UInt32 type,
-        OSDictionary* properties) override;
-    
-    virtual IOReturn externalMethod(
-        uint32_t selector,
-        IOExternalMethodArguments* arguments,
-        IOExternalMethodDispatch* dispatch = nullptr,
-        OSObject* target = nullptr,
-        void* reference = nullptr) override;
-    virtual IOReturn clientMemoryForType(
-        UInt32 type,
-        IOOptionBits* options,
-        IOMemoryDescriptor** memory) override;
-    virtual IOReturn registerNotificationPort(
-        mach_port_t port,
-        UInt32 type,
-        io_user_reference_t refCon) override;
-    
-    virtual IOReturn clientClose() override;
-
-    // IOService methods:
-    virtual void stop(IOService* provider) override;
-
-    // OSObject methods:
-    virtual void free() override;
-
-
-    void sendMessage(const void* message, uint32_t size);
-
-private:
-    void cleanupProviderRegistration();
-
-    // External methods:
-    static IOReturn registerVirtualizationRoot(
-        OSObject* target,
-        void* reference,
-        IOExternalMethodArguments* arguments);
-    IOReturn registerVirtualizationRoot(const char* rootPath, size_t rootPathSize, uint64_t* outError);
-
-    static IOReturn kernelMessageResponse(
-        OSObject* target,
-        void* reference,
-        IOExternalMethodArguments* arguments);
-    IOReturn kernelMessageResponse(uint64_t messageId, MessageType responseType);
-};
+void ProviderUserClient_UpdatePathProperty(PrjFSProviderUserClient* userClient, const char* providerPath);
+void ProviderUserClient_Retain(PrjFSProviderUserClient* userClient);
+void ProviderUserClient_Release(PrjFSProviderUserClient* userClient);
+void ProviderUserClient_SendMessage(PrjFSProviderUserClient* userClient, const void* message, uint32_t size);
```

Comment:
This means we need to use these wrapper functions to interact with opaque `PrjFSProviderUserClient` object pointers in all testable areas of the kext.

Reply:
Makes sense!  Wrappers for testing.
For my own understanding would mocks have been an option here or would it have been too involved/messy.

Reply:
Doing anything else would have caused issues due to the way the testable kext target is compiled - that one pulls in the real kernel headers, so it would really have been an `IOUserClient` subclass. `IOUserClient` (inheritance chain: `IOUserClient : IOService : IORegistryEntry : OSObject : OSMetaClassBase`) literally has dozens of fields and even more member functions. Because the testable kext is built against the kernel headers, this means the in-memory layouts of the class itself and the vtable are all fixed from those. So any mock implementations would have to match those layouts 1:1. We'd have to provide dummy implementations for all those member functions, etc. Just far more effort than it's worth.